In [25]:
import timm

from timm.data import resolve_data_config
from timm.data.transforms_factory import create_transform
model = timm.create_model('tf_efficientnetv2_b3', num_classes=2)
config = resolve_data_config({}, model=model)
#transform = create_transform(**config)

In [26]:
config

{'input_size': (3, 240, 240),
 'interpolation': 'bicubic',
 'mean': (0.485, 0.456, 0.406),
 'std': (0.229, 0.224, 0.225),
 'crop_pct': 0.904}

In [1]:
import random

import numpy as np
import pandas as pd

from dataset import get_dataset
from trainer import Trainer
from inference import create_label, sum_label
from loss import LabelSmoothing

import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader
from torchvision import models

In [2]:
## seeds

random_seed = 42

torch.manual_seed(random_seed)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
np.random.seed(random_seed)
random.seed(random_seed)

In [3]:
## hyperparameters

model_name = 'efficientnet-b0'
batch_size = 64
lr = 0.00006
num_epoch = 30
target = 'gender'
split = 20

df_train_path = f'/opt/ml/code/df/df_mask_gender_train_{split}.csv'
df_valid_path = f'/opt/ml/code/df/df_mask_gender_valid_{split}.csv'
df_test_path = '/opt/ml/input/data/eval/info.csv'

In [4]:
## prepare dataset

df_train = pd.read_csv(df_train_path)
df_valid = pd.read_csv(df_valid_path)
df_test = pd.read_csv(df_test_path)

train_dataset, valid_dataset, test_dataset = get_dataset(df_train, df_valid, df_test, target=target)

In [5]:
## prepare sampler

from torch.utils.data.sampler import WeightedRandomSampler

target = train_dataset.target
weight = df_train[target].value_counts().sort_index().to_numpy()
print('count :', weight)
weight = 1. / weight
samples_weight = np.array([weight[t] for t in df_train[target]])
samples_weight = torch.from_numpy(samples_weight)
sampler = WeightedRandomSampler(samples_weight.type('torch.DoubleTensor'), len(samples_weight)//2)

count : [5831 9289]


In [6]:
## prepare Dataloader

dataloaders = {'train' : DataLoader(train_dataset, batch_size=batch_size, num_workers=3, drop_last=True, sampler=sampler),
               'valid' : DataLoader(valid_dataset, batch_size=batch_size, shuffle=False, num_workers=3, drop_last=False)}
test_dataloader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False, num_workers=3, drop_last=False)

In [ ]:
# check input train batch
a = iter(dataloaders['valid'])
from utils import imshow
for i in range(4):
    batch = next(a)
imshow(batch[0])
print(batch[1])

In [7]:
## prepare model

device = 'cuda'

if model_name.startswith('efficientnet'):
    from efficientnet_pytorch import EfficientNet
    model = EfficientNet.from_pretrained(model_name, num_classes=len(train_dataset.classes))
else:
    model = models.resnet18(pretrained=True)
    model.fc = nn.Linear(model.fc.in_features, len(train_dataset.classes))
    
model.to(device)
print(model_name, 'ready')

Loaded pretrained weights for efficientnet-b0
efficientnet-b0 ready


In [8]:
criterion = LabelSmoothing(0.05)
optimizer = optim.Adam(model.parameters(), lr=lr)
lr_scheduler = None #optim.lr_scheduler.MultiStepLR(optimizer, milestones=[5, 10, 15], gamma = 0.2)

## Train

In [9]:
trainer = Trainer('/opt/ml/code/save')

In [10]:
## resume
# model.load_state_dict(torch.load('/opt/ml/code/save/label/effnet_test018.pt'))

# # change lr manually:
# for g in optimizer.param_groups:
#     g['lr'] = 0.00001

In [11]:
trainer.train(model, dataloaders, criterion, optimizer, device, num_epochs=num_epoch, scheduler=lr_scheduler,
              sub_dir='gender', save_name='epoch_')

Epoch 1/30
----------


100%|██████████| 118/118 [00:32<00:00,  3.60it/s]


train Loss: 0.4217 Acc: 0.8550 F1: 0.8549256048647949


100%|██████████| 60/60 [00:07<00:00,  7.62it/s]


valid Loss: 0.2203 Acc: 0.9643 F1: 0.9626845780691935

Epoch 2/30
----------


100%|██████████| 118/118 [00:33<00:00,  3.51it/s]


train Loss: 0.2282 Acc: 0.9510 F1: 0.951004032965106


100%|██████████| 60/60 [00:07<00:00,  7.65it/s]


valid Loss: 0.1811 Acc: 0.9730 F1: 0.9716764824888728

Epoch 3/30
----------


100%|██████████| 118/118 [00:32<00:00,  3.61it/s]


train Loss: 0.2011 Acc: 0.9629 F1: 0.9629230631115899


100%|██████████| 60/60 [00:07<00:00,  7.71it/s]


valid Loss: 0.1745 Acc: 0.9783 F1: 0.9770816333351178

Epoch 4/30
----------


100%|██████████| 118/118 [00:32<00:00,  3.58it/s]


train Loss: 0.1906 Acc: 0.9677 F1: 0.967688500173421


100%|██████████| 60/60 [00:07<00:00,  7.58it/s]


valid Loss: 0.1658 Acc: 0.9820 F1: 0.9810333881434768

Epoch 5/30
----------


100%|██████████| 118/118 [00:32<00:00,  3.59it/s]


train Loss: 0.1771 Acc: 0.9744 F1: 0.9744379284840536


100%|██████████| 60/60 [00:07<00:00,  7.65it/s]


valid Loss: 0.1672 Acc: 0.9820 F1: 0.9810189172807606

Epoch 6/30
----------


100%|██████████| 118/118 [00:33<00:00,  3.55it/s]


train Loss: 0.1695 Acc: 0.9796 F1: 0.9796050241090456


100%|██████████| 60/60 [00:07<00:00,  7.70it/s]


valid Loss: 0.1663 Acc: 0.9802 F1: 0.9790407866292196

Epoch 7/30
----------


100%|██████████| 118/118 [00:32<00:00,  3.60it/s]


train Loss: 0.1660 Acc: 0.9789 F1: 0.9789457792903742


100%|██████████| 60/60 [00:07<00:00,  7.62it/s]


valid Loss: 0.1597 Acc: 0.9823 F1: 0.9812860764522078

Epoch 8/30
----------


100%|██████████| 118/118 [00:32<00:00,  3.62it/s]


train Loss: 0.1615 Acc: 0.9825 F1: 0.9825195975557064


100%|██████████| 60/60 [00:07<00:00,  7.66it/s]


valid Loss: 0.1616 Acc: 0.9807 F1: 0.979594424960853

Epoch 9/30
----------


100%|██████████| 118/118 [00:33<00:00,  3.56it/s]


train Loss: 0.1603 Acc: 0.9823 F1: 0.9822549172267416


100%|██████████| 60/60 [00:07<00:00,  7.68it/s]


valid Loss: 0.1638 Acc: 0.9815 F1: 0.9804456268706352

Epoch 10/30
----------


100%|██████████| 118/118 [00:32<00:00,  3.61it/s]


train Loss: 0.1554 Acc: 0.9841 F1: 0.9841100090110613


100%|██████████| 60/60 [00:07<00:00,  7.72it/s]


valid Loss: 0.1568 Acc: 0.9839 F1: 0.9829965670881589

Epoch 11/30
----------


 43%|████▎     | 51/118 [00:14<00:19,  3.45it/s]


KeyboardInterrupt: 

## Experiment Logs

---

## Gender Experiment Logs

---

1. model=eff b0, optimizer=Adam, lr=0.00006, bs=64, augment=randaug+cutout, inputsize=224, loss_fn=LS(0.05), split 20%
    * epoch 18)
      - train Loss: 0.1446 Acc: 0.9881 F1: 0.9880777974235919
      - valid Loss: 0.1551 Acc: 0.9854 F1: 0.9846651730146876
        
2. model=eff b0, optimizer=Adam, lr=0.0003, bs=64, augment=randaug+cutout, inputsize=224, loss_fn=LS(0.05), split 20%
    * epoch 15)
      - train Loss: 0.1371 Acc: 0.9911 F1: 0.9911258239235885
      - valid Loss: 0.1574 Acc: 0.9868 F1: 0.9860680087572518
---
3. model=eff b0, optimizer=Adam, lr=0.0003, bs=64, augment=randaug+cutout, inputsize=224, crop=224, loss_fn=LS(0.05), split 20%
    * epoch 09)
      - train Loss: 0.1574 Acc: 0.9799 F1: 0.9798725637365424
      - valid Loss: 0.1556 Acc: 0.9820 F1: 0.9810713788113145
---
4. model=eff b0, optimizer=Adam, lr=0.0003, bs=64, augment=randaug+cutout, inputsize=224, crop=280, loss_fn=LS(0.05), split 20%
    * epoch 09)
      - train Loss: 0.1574 Acc: 0.9799 F1: 0.9798725637365424
      - valid Loss: 0.1556 Acc: 0.9820 F1: 0.9810713788113145

### loss function test

---

1. eff b0, lr=0.00006, bs=64, randaug+cutout, inputsize=224, loss_fn=CE, WeightedSampler, split 25%, optimizer=Adam
   * epoch 18)
     - train Loss: 0.0991 Acc: 0.9655 F1: 0.9654261306812848
     - valid Loss: 0.0980 Acc: 0.9778 F1: 0.9772288369633227  
  * before first patience

---

2. eff b0, lr=0.00006, bs=64, randaug+cutout, inputsize=224, loss_fn=LS(0.05), WeightedSampler, split 25%, optimizer=Adam
    * epoch 21)
     - train Loss: 0.3310 Acc: 0.9638 F1: 0.963728513759618
     - valid Loss: 0.3071 Acc: 0.9786 F1: 0.9791634664746235 

---

3. eff b0, lr=0.00006, bs=64, randaug+cutout, inputsize=224, loss_fn=F1_Loss, split 25%, optimizer=Adam
   * epoch 13)
     - train Loss: 0.0710 Acc: 0.9584 F1: 0.9476713834498373
     - valid Loss: 0.0403 Acc: 0.9765 F1: 0.9748789208889367
   * epoch 3)
     - train Loss: 0.0594 Acc: 0.9655 F1: 0.9555452013464937
     - valid Loss: 0.0369 Acc: 0.9782 F1: 0.9776768545260203

---

### split ratio test

---

4. model=eff b0, optimizer=Adam, lr=0.00006, bs=64, augment=randaug+cutout, inputsize=224, loss_fn=LS(0.05), split 15%
    * epoch 21)
      - train Loss: 0.3381 Acc: 0.9597 F1: 0.9597144587078792
      - valid Loss: 0.2948 Acc: 0.9852 F1: 0.982564296238198
      
---

5. model=eff b0, optimizer=Adam, lr=0.00006, bs=64, augment=randaug+cutout, inputsize=224, loss_fn=LS(0.05), split 20%
    * epoch 27)
      - train Loss: 0.3211 Acc: 0.9668 F1: 0.9668728849468181
      - valid Loss: 0.2878 Acc: 0.9854 F1: 0.9842203787970396
      
---

6. model=eff b0, optimizer=Adam, lr=0.00006, bs=64, augment=randaug+cutout, inputsize=224, loss_fn=LS(0.05), split 25%
    * epoch 22)
      - train Loss: 0.3321 Acc: 0.9639 F1: 0.9638739223128628
      - valid Loss: 0.2910 Acc: 0.9829 F1: 0.9792575649184908
      
---

7. model=eff b0, optimizer=Adam, lr=0.00006, bs=64, augment=randaug+cutout, inputsize=224, loss_fn=LS(0.05), split 30%
    * epoch 29)
      - train Loss: 0.3171 Acc: 0.9686 F1: 0.9685813478405522
      - valid Loss: 0.3047 Acc: 0.9772 F1: 0.9764751322879572

## Debug

In [ ]:
from sklearn.metrics import plot_confusion_matrix
from sklearn.metrics import confusion_matrix

model.eval()
y_true = []
y_pred = []
wrong = []
for inputs, labels in dataloaders['valid']:
    inputs = inputs.to(device)
    labels = labels.to(device)

    with torch.no_grad():
        outputs = model(inputs)
        _, preds = torch.max(outputs, 1)
        y_true.extend(labels.tolist())
        y_pred.extend(preds.tolist())
        
        for p, t, inp in zip(preds, labels, inputs):
            if p.item() != t.item():
                wrong.append((inp.cpu().numpy(), p.item(), t.item()))

mtx = confusion_matrix(y_true, y_pred)
print(mtx)


In [ ]:
import matplotlib.pyplot as plt

label_ = list(range(6))

fig, ax = plt.subplots(figsize=(8,8))
im = ax.imshow(mtx, cmap='Blues', vmax=100, vmin=-20)

ax.set_xticks(np.arange(len(label_)))
ax.set_yticks(np.arange(len(label_)))
# ... and label them with the respective list entries
ax.set_xticklabels(label_)
ax.set_yticklabels(label_)

# Rotate the tick labels and set their alignment.
plt.setp(ax.get_xticklabels(), rotation=45, ha="right",
         rotation_mode="anchor")

# Loop over data dimensions and create text annotations.
for i in range(len(label_)):
    for j in range(len(label_)):
        if mtx[i, j]:
            text = ax.text(j, i, mtx[i, j],
                           ha="center", va="center", color="w", size=10)
ax.set_xlabel('Prediction')
ax.set_ylabel('True Label')
ax.set_title("Confusion mtx")
fig.tight_layout()
plt.show()

In [ ]:
import matplotlib.pyplot as plt
iterator = iter(wrong)

label = ["Male with mask", "Female with Mask", "Male incorrect mask", "Female no mask", "Male incorrect mask", "Female no mask"]
mean = np.array([0.485, 0.456, 0.406])
std = np.array([0.229, 0.224, 0.225])

In [ ]:
fig, axes = plt.subplots(2, 2, figsize=(10,10))
for ax in axes.flatten():
    img, p, t = next(iterator)
    img = img.transpose((1, 2, 0))
    img = std * img + mean
    img = np.clip(img, 0, 1)
    ax.imshow(img)
    ax.axis('off')
    ax.set_title(f't : {label[t]}\np : {label[p]}')
plt.show()

## Inference

In [ ]:
def sum_label(df):
    
    def _label(row):
        return row['age'] + 3*row['label']# + 6*row['mask']

    df['ans'] = df.apply(_label, axis=1)
    return df[['ImageID', 'ans']]

In [ ]:
## inference

# model.load_state_dict(torch.load('/opt/ml/code/save/best_log2.pt'))
# df_submit = create_label(model, test_dataloader, df_test.copy(), device, target='label')

# model = EfficientNet.from_pretrained(model_name, num_classes=3).to(device)
# model.load_state_dict(torch.load('/opt/ml/code/save/age/extended_far_best.pt'))
# df_submit = create_label(model, test_dataloader, df_submit, device, target='age')

# model.load_state_dict(torch.load('/opt/ml/code/save/mask/6e5_9987.pt'))
# df_submit = create_label(model, test_dataloader, df_submit, device, target='mask')

# model.fc = nn.Linear(model.fc.in_features, 2).cuda()
# model.load_state_dict(torch.load('/opt/ml/code/save/gender/6e5_9851.pt'))
# df_submit = create_label(model, test_dataloader, df_submit, device, target='gender')


# df_submit = sum_label(df_submit)
# df_submit


df_submit.to_csv('submit.csv', index=False)